{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e0e105f7-efaf-4dba-b2fc-98feccebfe48",
   "metadata": {},
   "source": [
    "# Create ground truth estiamtions of PSI values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b151eda-974e-4aad-b61a-0e3c02f6bed7",
   "metadata": {
    "tags": []
   },
   "source": [
    "Code adapted from `simulate-batcheffect.r` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cab3a2c9-c117-43d1-a779-d8282addfce1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "'drc' has been loaded.\n",
      "\n",
      "\n",
      "Please cite R and 'drc' if used for a publication,\n",
      "\n",
      "for references type 'citation()' and 'citation('drc')'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘drc’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:VGAM’:\n",
      "\n",
      "    gompertz, logistic\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    gaussian, getInitial\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:VGAM’:\n",
      "\n",
      "    fill\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(readr)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(VGAM)\n",
    "library(drc)\n",
    "library(tidyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9859f1e2-86e3-4687-9ca8-91e80bd37e23",
   "metadata": {
    "tags": []
   },
   "source": [
    "## TRex Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e91079e-3068-4c9c-a817-60be9042fd59",
   "metadata": {},
   "outputs": [],
   "source": [
    "trex_counts<-function(counts,trex.map,...){\n",
    "  \n",
    "    \n",
    " trex_sum <- as.data.frame(counts) %>%\n",
    "                tibble::rownames_to_column(\"ensembl_transcript_id\") %>%\n",
    "                inner_join(.,trex.map,by = \"ensembl_transcript_id\") %>%\n",
    "                dplyr::select(-ensembl_transcript_id) %>%\n",
    "                group_by(event_id) %>%\n",
    "                summarize_all(sum)\n",
    "    \n",
    "  return(trex_sum)\n",
    "}\n",
    "\n",
    "trex_means<-function(lengths,trex.map,...){\n",
    "  \n",
    "  trex_mean<-aggregate(lengths[trex.map$ensembl_transcript_id,],\n",
    "                       list(event_name = trex.map$event_name),\n",
    "                       function(x){mean(x,na.rm = F)}) %>% \n",
    "    tibble::column_to_rownames(\"event_name\")\n",
    "  return(trex_mean)\n",
    "}\n",
    "\n",
    "correct_sim_counts<-function(trex){\n",
    "\n",
    "    mats<-list()\n",
    "    for(tc in c(\"A\",\"C\")){\n",
    "        cols <- grepl(paste0(tc,\"_\"),colnames(trex$counts))\n",
    "\n",
    "        # Retrieve counts and length matrices\n",
    "        countsMat <- trex$counts[,cols]\n",
    "        lengthMat <- trex$length[,cols] \n",
    "        tpmMat <- trex$abundance[,cols]\n",
    "\n",
    "        # Re-esimate abundances for the trex count type\n",
    "        tpmSum <- colSums(tpmMat)\n",
    "        abundanceMat <- t(t(tpmMat)/tpmSum*1e6)\n",
    "\n",
    "        # Estimate counts from abundances per trex count type\n",
    "        eCountsMat <- abundanceMat * rowMeans(lengthMat) # Scale to average length over all samples\n",
    "        scaleFactor <- colSums(countsMat,na.rm=T)/colSums(eCountsMat,na.rm=T)\n",
    "        eCountsMat <- t(t(eCountsMat) * scaleFactor)\n",
    "\n",
    "        mats[[tc]] <- list(\"counts\"=eCountsMat,\"abundance\"=abundanceMat,\"length\"=lengthMat)\n",
    "    }\n",
    "    trex$counts<-cbind(mats[[\"A\"]][[\"counts\"]],mats[[\"C\"]][[\"counts\"]])\n",
    "    trex$abundance<-cbind(mats[[\"A\"]][[\"abundance\"]],mats[[\"C\"]][[\"abundance\"]])\n",
    "    trex$length<-cbind(mats[[\"A\"]][[\"length\"]],mats[[\"C\"]][[\"length\"]])\n",
    "    trex$countsFromAbundance<-\"lengthScaledTPM\"\n",
    "\n",
    "    return(trex)\n",
    "}\n",
    "\n",
    "create_trex_object<-function(transcript.txi,alt_isos,const_isos){\n",
    "\n",
    "    message(\"Converting to event counts...\")\n",
    "    \n",
    "    tpm<-transcript.txi$abundance\n",
    "    countsMat<-transcript.txi$counts\n",
    "    lengthMat<-transcript.txi$length\n",
    "    \n",
    "    tpm_alt<-trex_counts(tpm,alt_isos) \n",
    "    colnames(tpm_alt)[-1]<-paste0(\"A_\",colnames(tpm_alt)[-1])\n",
    "\n",
    "    tpm_const<-trex_counts(tpm,const_isos)\n",
    "    colnames(tpm_const)[-1]<-paste0(\"C_\",colnames(tpm_const)[-1])\n",
    "\n",
    "    cts_alt<-trex_counts(countsMat,alt_isos)\n",
    "    colnames(cts_alt)[-1]<-paste0(\"A_\",colnames(cts_alt)[-1])\n",
    "\n",
    "    cts_const<-trex_counts(countsMat,const_isos) \n",
    "    colnames(cts_const)[-1]<-paste0(\"C_\",colnames(cts_const)[-1]) \n",
    "    \n",
    "    len_alt<-inner_join(alt_isos,lengthMat %>% tibble::rownames_to_column(\"ensembl_transcript_id\"),by = \"ensembl_transcript_id\") %>%\n",
    "                group_by(event_id) %>% dplyr::select(-ensembl_transcript_id) %>%\n",
    "                summarize_all(mean)\n",
    "    colnames(len_alt)[-1]<-paste0(\"A_\",colnames(len_alt)[-1])\n",
    "\n",
    "    len_const<-inner_join(const_isos,lengthMat %>% tibble::rownames_to_column(\"ensembl_transcript_id\"),by = \"ensembl_transcript_id\") %>%\n",
    "                     group_by(event_id) %>% dplyr::select(-ensembl_transcript_id) %>%\n",
    "                    summarize_all(mean)\n",
    "    colnames(len_const)[-1]<-paste0(\"C_\",colnames(len_const)[-1])\n",
    "\n",
    "    trex<-list()\n",
    "    trex$abundance<-inner_join(tpm_alt,tpm_const,by=\"event_id\") %>% tibble::column_to_rownames(\"event_id\")\n",
    "    trex$counts<-inner_join(cts_alt,cts_const,by=\"event_id\") %>% tibble::column_to_rownames(\"event_id\")\n",
    "    trex$length<-inner_join(len_alt,len_const,by=\"event_id\") %>% tibble::column_to_rownames(\"event_id\")\n",
    "    \n",
    "    trex<-correct_sim_counts(trex)\n",
    "    \n",
    "    return(trex)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69a35cc-461b-4bc9-9a32-61f13e1b79f6",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "93f7c097-c838-43e8-b6e7-66f26570ab67",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_input <- \"/project/6007998/lmoral7/altsplicing-methods/simulations/data/observed_counts\"\n",
    "effect_dir <- \"/project/6007998/lmoral7/altsplicing-methods/simulations/data/sampled_counts/gt_effects\"\n",
    "asmapfile <- \"/home/lmoral7/lmprojects/references/gencode_GRCh38.p13/SUPPA2_splicemap/modID.GRCh38.p13.gencode.v37.primary_assembly.annotation_ASALL_strict.ioe\"\n",
    "sample_file <- \"K562_control_rep1/rsem.isoforms.results\"\n",
    "sample_file_2 <- \"K562_SRSF9.KD_rep1/rsem.isoforms.results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e70d44ec-7053-4f32-985f-f10921a8f38e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Splicing GTF map\n",
    "asmap<-fread(asmapfile,data.table = F,stringsAsFactors = F) %>% \n",
    "        as.data.frame() %>% \n",
    "        filter(event_type==\"SE\") %>% distinct() \n",
    "rownames(asmap)<-asmap$event_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "43c08d56-40e2-4ae2-8828-30e6d55f75da",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ground truth counts \n",
    "\n",
    "iso_res_rsem<-read.table(file.path(dir_input,sample_file),header=TRUE)\n",
    "iso_res_kd_rsem<-read.table(file.path(dir_input,sample_file_2),header=TRUE)\n",
    "obs_TPM <- fread(file.path(dir_input,\"ground_truth_TPM.tsv\"),data.table=FALSE)\n",
    "simulation_design <- fread(file.path(dir_input,\"metadata.tsv\"))\n",
    "\n",
    "n_transcripts <- nrow(obs_TPM)\n",
    "pseudocount <- 1e-3\n",
    "set.seed(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3940d261-e0c8-4e57-88da-29a0842299af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining, by = \"transcript_id\"\n",
      "\u001b[1m\u001b[22mJoining, by = \"transcript_id\"\n"
     ]
    }
   ],
   "source": [
    "nonZero_TPMs <- rowSums( obs_TPM > 0 ) \n",
    "lengthMat <-  inner_join(iso_res_rsem[,c(\"transcript_id\",\"effective_length\")]  %>% rename(\"K562_control\"=\"effective_length\") ,\n",
    "                         iso_res_kd_rsem[,c(\"transcript_id\",\"effective_length\")] %>% rename(\"K562_SRSF9.KD\"=\"effective_length\")) %>%\n",
    "              tibble::column_to_rownames(\"transcript_id\")\n",
    "countsMat <- inner_join(iso_res_rsem[,c(\"transcript_id\",\"expected_count\")]  %>% rename(\"K562_control\"=\"expected_count\") ,\n",
    "                         iso_res_kd_rsem[,c(\"transcript_id\",\"expected_count\")] %>% rename(\"K562_SRSF9.KD\"=\"expected_count\")) %>%\n",
    "              tibble::column_to_rownames(\"transcript_id\")\n",
    "\n",
    "lengthMat<-cbind(lengthMat,lengthMat,lengthMat,lengthMat,lengthMat)\n",
    "colnames(lengthMat)<-paste(colnames(lengthMat),rep(paste0(\"rep\",1:5),each = 2),sep=\"_\")\n",
    "lengthMat<-lengthMat[,simulation_design$Sample]\n",
    "\n",
    "countsMat<-cbind(countsMat,countsMat,countsMat,countsMat,countsMat)\n",
    "colnames(countsMat)<-paste(colnames(countsMat),rep(paste0(\"rep\",1:5),each = 2),sep=\"_\")\n",
    "countsMat<-countsMat[,simulation_design$Sample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57e938dd-43b2-4910-9829-8d235d387ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ground truth effect sizes \n",
    "pars <- list.files(effect_dir)\n",
    "gt_effects <- lapply(pars, function(p,...){\n",
    "    ref<-read.table(file = file.path(effect_dir,p),sep = \"\\t\")\n",
    "})\n",
    "names(gt_effects)<-sub(\"_ground_truth_effects.tsv\",\"\",pars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3eb200ae-a336-4d06-9b7b-76115df75ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assign events to isoforms \n",
    "alt_isos<-asmap %>%\n",
    "        group_by(event_id)%>%\n",
    "        mutate(ensembl_transcript_id=strsplit(alternative_transcripts,\",\")) %>%\n",
    "        tidyr::unnest(ensembl_transcript_id) %>%\n",
    "        dplyr::select(event_id,ensembl_transcript_id) %>% ungroup()\n",
    "\n",
    "const_isos<-asmap %>%\n",
    "            group_by(event_id)%>%\n",
    "            mutate(ensembl_transcript_id=strsplit(constitutive_transcripts,\",\")) %>%\n",
    "            tidyr::unnest(ensembl_transcript_id) %>%\n",
    "            dplyr::select(event_id,ensembl_transcript_id) %>% ungroup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60b422af-a78b-4eef-83ad-dce59a5ab3da",
   "metadata": {},
   "source": [
    "## Ground truth effects "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "520975cb-4bb9-4ea5-b735-f312a24d9fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate simulated log TPM\n",
    "pseudocount<-1e-6\n",
    "design_mx <- model.matrix( ~ cell_line + batch, simulation_design )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e6074b98-7eac-4510-afb6-055f3d984d81",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt.effects_to_gt.PSI<-function(gt_effects_mat,...){\n",
    "    \n",
    "    message(\"Generating logTPM...\")\n",
    "    gt_effects_mat <- as.matrix(gt_effects_mat)\n",
    "    logTPM <- gt_effects_mat %*% t(design_mx)\n",
    "    tpm<-exp(logTPM)\n",
    "    tpm[nonZero_TPMs==0,] <- 0\n",
    "    tpm <- scale(tpm, center = F, scale=colSums(tpm)/1000000)\n",
    "    colnames(tpm)<-simulation_design$Sample\n",
    "    \n",
    "    # Get event counts \n",
    "    txi<-list(\"abundance\"=tpm,\n",
    "              \"counts\"=countsMat,\n",
    "             \"length\"=lengthMat)\n",
    "    trex<-create_trex_object(txi,alt_isos,const_isos) \n",
    "    \n",
    "    message(\"Regressing out effect sizes on PSI\")\n",
    "    tpm_alt<-trex$abundance[,grepl(\"A_\",colnames(trex$abundance))]\n",
    "    tpm_const<-trex$abundance[,grepl(\"C_\",colnames(trex$abundance))]\n",
    "    \n",
    "    psi<-tpm_alt/(tpm_alt+tpm_const)\n",
    "    psi<-t(log(psi/(1-psi))) #change to logit psi\n",
    "    psi[is.na(psi)|is.infinite(psi)]<-log(pseudocount/(1-pseudocount))\n",
    "    \n",
    "    psi.fit<-lm(psi ~ cell_line + batch,data = simulation_design)\n",
    "    \n",
    "    return(t(psi.fit$coefficients))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "78294c16-4a06-4a2c-b17d-b0f2f720df9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n",
      "Generating logTPM...\n",
      "\n",
      "Converting to event counts...\n",
      "\n",
      "Regressing out effect sizes on PSI\n",
      "\n"
     ]
    }
   ],
   "source": [
    "psi_effects<-lapply(gt_effects,gt.effects_to_gt.PSI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d0ab291-0107-40ba-bd3d-eff1e765f4a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(psi_effects,file = \"../data/ground_truth_effect_sizes_on_lenCorrected_logitPSI.RData\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
