{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41b642e6-2770-43e6-a7fb-eefd45e95eca",
   "metadata": {},
   "source": [
    "# Functional analysis of relevant RBPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "30c2c749-595a-424f-98cd-6c5f5355c2f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "Loading required package: future\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.8.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(furrr)\n",
    "library(purrr)\n",
    "library(dplyr)\n",
    "library(ggVennDiagram)\n",
    "library(ggplot2)\n",
    "library(ggsci)\n",
    "library(fgsea)\n",
    "library(ComplexHeatmap)\n",
    "library(matrixStats)\n",
    "library(ggpubr)\n",
    "library(msigdbr)\n",
    "library(patchwork)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "807c9e93-daa5-47bc-bad0-68cfae15e866",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6f022dd4-b23b-4cfe-8b64-0b7e71c2d7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Process AffiMx output\n",
    "#------------------------------------\n",
    "process_file <- function(file_path) {\n",
    "    \n",
    "  data <- fread(file_path, header = FALSE, sep = \"\\t\",data.table=FALSE) %>%\n",
    "            mutate(motif_id=sub(\"0\\\\.6_.*\",\"0.6\",basename(file_path)))\n",
    "  colnames(data)[1:2]<-c(\"Name\",\"affinity\")\n",
    "  data <- data %>%\n",
    "          mutate(event_id = sub(\".*ENS\",\"ENS\",Name) %>% sub(\"CHR\",\"chr\",.),\n",
    "                 seq_id = sub(\"_ENS.*\",\"\",Name) %>% sub(\"^([A-Z0-9]+)_\",\"\",.) %>% sub(\"_([0-9]*|\\\\d+\\\\.\\\\d+E\\\\+\\\\d+)_\",\"_\",.))\n",
    "  data<-data\n",
    "    \n",
    " return(data)\n",
    "}\n",
    "\n",
    "## Process all files \n",
    "#------------------------------------\n",
    "process_all_motif_files <- function(directory_path, motif_ids, window_size, event_type) {\n",
    "    \n",
    "  files <- file.path(directory_path, paste0(motif_ids,\"_spliceSites_windowSize_\",window_size,\"_up_and_down.affinity_\",event_type,\".txt\"))  \n",
    "  mdata <- furrr::future_map_dfr(files, process_file) %>%\n",
    "                      mutate(window_size=window_size,\n",
    "                             event_type=event_type)  \n",
    "  return(mdata)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8ca31f4-3cc9-438b-a815-831884e29f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wrapper to run the function that loads the affinity files and the function that runs the tests of targets\n",
    "find_targets_all_motifs<-function(motif_ids,event_type,window_size,dir,trex,...){\n",
    "    \n",
    "    rbp.aff <- process_all_motif_files(directory_path = dir,\n",
    "                                       motif_ids = motif_ids,\n",
    "                                       event_type = event_type,\n",
    "                                       window_size = window_size)\n",
    "    # Filter to test only biologically revelant genes\n",
    "    rbp.aff <- rbp.aff %>%\n",
    "               filter(event_id %in% unique(trex$event_id))\n",
    "    \n",
    "    rbp.targets <- test_motif_targets(rbp.aff,p=0.15) %>%\n",
    "                   mutate(window_size = window_size)\n",
    "    \n",
    "    return(rbp.targets)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b8dc3add-f4af-4ba4-b960-f394784e4f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test for target genes\n",
    "test_motif_targets <- function(rbp.aff,p){\n",
    "\n",
    "    target.tests <- rbp.aff %>%\n",
    "                    mutate(gene_id = sub(\".\\\\d+;.*\",\"\",event_id)) %>%       \n",
    "                    group_by(event_type,window_size,seq_id,motif_id,gene_id) %>% \n",
    "                    summarize(gene_affinity = mean(log(affinity)),\n",
    "                              nevents=length(unique(event_id))) %>%   # Summarize is as mean affinity per region over all events in the same gene\n",
    "                    group_by(event_type,window_size,seq_id,motif_id) %>% \n",
    "                    mutate(ma = mean(gene_affinity),\n",
    "                           zscore = (gene_affinity-ma)/sd(gene_affinity),\n",
    "                           n = length(zscore),\n",
    "                           zscore_pvalue = pnorm(zscore,lower.tail = TRUE),\n",
    "                           zscore_padj = p.adjust(zscore_pvalue,method=\"fdr\")) %>%\n",
    "                    rename(\"representative_motif\"=\"motif_id\") %>%\n",
    "                    ungroup() \n",
    "    \n",
    "    target.genes <- target.tests %>%\n",
    "                    mutate(target_gene_region = as.character(ifelse(zscore_padj<p,\"target\",\"other\"))) %>%\n",
    "                    arrange(zscore_padj)\n",
    "    \n",
    "    return(target.genes)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7505b42-e34c-42ad-ab94-1ac81bdf4a05",
   "metadata": {},
   "source": [
    "### Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "91c9f357-281e-4cef-8a76-2d59f2ee3ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pars\n",
    "analysis<-\"condition\"  # Analysis type tag (condition / TumorStageContinuous)\n",
    "voi<-\"conditiontumor\"  # Variable of interest (conditiontumor / sex / age / stage)\n",
    "events_used<-c(\"SE\",\"RI\",\"AF\",\"AL\",\"A5\",\"A3\") # Events to include in the analysis\n",
    "n<-11 # Number of cancers to call shared events\n",
    "set.seed(7)\n",
    "cancer_order_set<-c(\"SKCM\",\"UVM\",\"SARC\",\"LAML\",\"TGCT\",\"UCEC\",\"CESC\",\"PRAD\",\"READ\",\"BLCA\",\n",
    "                \"COAD\",\"PAAD\",\"CHOL\",\"KIRP\",\"KIRC\",\"KICH\",\"ACC\",\"PCPG\",\"STAD\",\"LIHC\",\n",
    "                \"MESO\",\"LUSC\",\"LUAD\",\"BRCA\",\"THCA\",\"ESCA\",\"HNSC\")\n",
    "\n",
    "# Event flag information\n",
    "obj_dir<-\"../input/objects\"\n",
    "flags_file<-paste0(obj_dir,\"/tcga.\",analysis,\".psi.flags.RDS\")\n",
    "\n",
    "# RBP info\n",
    "aff_dir<-\"../../../chapter2_Methods/2.5_ASregulators/output/affimx\"\n",
    "rbp_obj_file<-\"../output/objects/Figure10_rbp_motif_coefs_gex_sig.RDS\"\n",
    "rbp_cors_file<-\"../output/data/RBP_significant_correlations.tsv\"\n",
    "\n",
    "# Internal\n",
    "vars<-c(voi,\"impurity\")\n",
    "res.file<-paste0(\"../input/objects/tcga.\",analysis,\".res.lfcShrink.RDS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b7941ad-3083-40fb-9d19-67ff1458db4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plan(multisession, workers = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca326c3-6ce3-42d2-9c9c-aa7f630577ed",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "877f45bb-b0c4-4dc5-b6bb-ed4da57414a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_motif_coefs_gex_sig <- readRDS(file=rbp_obj_file)\n",
    "trex.res.lfsh <- readRDS(file = res.file) %>%\n",
    "               filter(event_type %in% events_used)\n",
    "flags <- readRDS(flags_file)\n",
    "signif.cors <- data.table::fread(rbp_cors_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49b83602-0b34-45ad-8941-1f42f0f2d96a",
   "metadata": {},
   "source": [
    "### Filter events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d6001184-0c56-40c8-b8c0-bc7582c40892",
   "metadata": {},
   "outputs": [],
   "source": [
    "trex.res.cond <- trex.res.lfsh %>%\n",
    "                 filter(exp_var==\"conditiontumor\",\n",
    "                       !is.na(padj)) %>%\n",
    "                 mutate(gene_id=sub(\"\\\\.\\\\d+\",\"\",gene_id))  %>%\n",
    "                 distinct(gene_id,event_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "842f098b-56a6-48b4-956f-0f47fe01acb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1m\u001b[22mJoining, by = \"event_id\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>213669</li><li>4</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 213669\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 213669\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 213669      4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "valid_events <- flags %>%\n",
    "                group_by(event_type,event_id) %>%\n",
    "                summarize(ncancers=length(unique(cancer))) %>%\n",
    "                filter(n!=20)\n",
    "trex.res.cond <- trex.res.cond%>%\n",
    "                inner_join(.,valid_events)\n",
    "dim(trex.res.cond)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef774aaf-7908-40d0-944f-d872fef04ea7",
   "metadata": {},
   "source": [
    "### Find targets of RBPs with significant correlations between LFC and DSR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7079f2ff-3352-481f-adc2-07b8c4567746",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig.rbp.motifs <- rbp_motif_coefs_gex_sig %>% \n",
    "                  ungroup() %>%\n",
    "                  distinct(representative_motif,event_type,wsize) %>%\n",
    "                  group_by(representative_motif,event_type,wsize) %>%\n",
    "                  mutate(pars=list(list(representative_motif,event_type,wsize))) %>%\n",
    "                  ungroup() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "221459d4-df45-46a8-84bc-332af0d44482",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'event_type', 'window_size', 'seq_id',\n",
      "'motif_id'. You can override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "motif_targets <- future_map_dfr(sig.rbp.motifs$pars,function(p,...){\n",
    "                                                    find_targets_all_motifs(motif_ids = p[[1]],\n",
    "                                                                            event_type = p[[2]],\n",
    "                                                                            window_size = p[[3]],\n",
    "                                                                            dir = aff_dir,\n",
    "                                                                            trex = trex.res.cond)}) %>%\n",
    "                distinct() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c63645ca-e1ca-439c-8b4e-0e56e5d423f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_targets_genes <- motif_targets %>%\n",
    "                       filter(target_gene_region==\"target\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6445f90b-d933-4de3-bd27-2760ed2d7bdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>event_type</th><th scope=col>window_size</th><th scope=col>seq_id</th><th scope=col>representative_motif</th><th scope=col>gene_id</th><th scope=col>gene_affinity</th><th scope=col>nevents</th><th scope=col>ma</th><th scope=col>zscore</th><th scope=col>n</th><th scope=col>zscore_pvalue</th><th scope=col>zscore_padj</th><th scope=col>target_gene_region</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000156959</td><td>-14.25651</td><td>1</td><td>-4.617702</td><td>-4.145691</td><td>8669</td><td>1.693954e-05</td><td>0.08531045</td><td>target</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S2_DOWN_200</td><td>M012_0.6</td><td>ENSG00000156959</td><td>-14.25651</td><td>1</td><td>-4.506304</td><td>-4.255960</td><td>8669</td><td>1.040770e-05</td><td>0.09022436</td><td>target</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S2_DOWN_200</td><td>M012_0.6</td><td>ENSG00000198858</td><td>-13.75724</td><td>1</td><td>-4.506304</td><td>-4.038029</td><td>8669</td><td>2.695104e-05</td><td>0.11681929</td><td>target</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000198858</td><td>-13.75724</td><td>1</td><td>-4.617702</td><td>-3.930953</td><td>8669</td><td>4.230485e-05</td><td>0.12224692</td><td>target</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>E1_DOWN_200</td><td>M012_0.6</td><td>ENSG00000243710</td><td>-13.56790</td><td>1</td><td>-4.196342</td><td>-4.057798</td><td>8669</td><td>2.476879e-05</td><td>0.13289494</td><td>target</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " event\\_type & window\\_size & seq\\_id & representative\\_motif & gene\\_id & gene\\_affinity & nevents & ma & zscore & n & zscore\\_pvalue & zscore\\_padj & target\\_gene\\_region\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000156959 & -14.25651 & 1 & -4.617702 & -4.145691 & 8669 & 1.693954e-05 & 0.08531045 & target\\\\\n",
       "\t A3 & 200 & S2\\_DOWN\\_200 & M012\\_0.6 & ENSG00000156959 & -14.25651 & 1 & -4.506304 & -4.255960 & 8669 & 1.040770e-05 & 0.09022436 & target\\\\\n",
       "\t A3 & 200 & S2\\_DOWN\\_200 & M012\\_0.6 & ENSG00000198858 & -13.75724 & 1 & -4.506304 & -4.038029 & 8669 & 2.695104e-05 & 0.11681929 & target\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000198858 & -13.75724 & 1 & -4.617702 & -3.930953 & 8669 & 4.230485e-05 & 0.12224692 & target\\\\\n",
       "\t A3 & 200 & E1\\_DOWN\\_200 & M012\\_0.6 & ENSG00000243710 & -13.56790 & 1 & -4.196342 & -4.057798 & 8669 & 2.476879e-05 & 0.13289494 & target\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 13\n",
       "\n",
       "| event_type &lt;chr&gt; | window_size &lt;chr&gt; | seq_id &lt;chr&gt; | representative_motif &lt;chr&gt; | gene_id &lt;chr&gt; | gene_affinity &lt;dbl&gt; | nevents &lt;int&gt; | ma &lt;dbl&gt; | zscore &lt;dbl&gt; | n &lt;int&gt; | zscore_pvalue &lt;dbl&gt; | zscore_padj &lt;dbl&gt; | target_gene_region &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000156959 | -14.25651 | 1 | -4.617702 | -4.145691 | 8669 | 1.693954e-05 | 0.08531045 | target |\n",
       "| A3 | 200 | S2_DOWN_200 | M012_0.6 | ENSG00000156959 | -14.25651 | 1 | -4.506304 | -4.255960 | 8669 | 1.040770e-05 | 0.09022436 | target |\n",
       "| A3 | 200 | S2_DOWN_200 | M012_0.6 | ENSG00000198858 | -13.75724 | 1 | -4.506304 | -4.038029 | 8669 | 2.695104e-05 | 0.11681929 | target |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000198858 | -13.75724 | 1 | -4.617702 | -3.930953 | 8669 | 4.230485e-05 | 0.12224692 | target |\n",
       "| A3 | 200 | E1_DOWN_200 | M012_0.6 | ENSG00000243710 | -13.56790 | 1 | -4.196342 | -4.057798 | 8669 | 2.476879e-05 | 0.13289494 | target |\n",
       "\n"
      ],
      "text/plain": [
       "  event_type window_size seq_id      representative_motif gene_id        \n",
       "1 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "2 A3         200         S3_DOWN_200 M012_0.6             ENSG00000156959\n",
       "3 A3         200         S2_DOWN_200 M012_0.6             ENSG00000156959\n",
       "4 A3         200         S2_DOWN_200 M012_0.6             ENSG00000198858\n",
       "5 A3         200         S3_DOWN_200 M012_0.6             ENSG00000198858\n",
       "6 A3         200         E1_DOWN_200 M012_0.6             ENSG00000243710\n",
       "  gene_affinity nevents ma        zscore    n    zscore_pvalue zscore_padj\n",
       "1 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "2 -14.25651     1       -4.617702 -4.145691 8669 1.693954e-05  0.08531045 \n",
       "3 -14.25651     1       -4.506304 -4.255960 8669 1.040770e-05  0.09022436 \n",
       "4 -13.75724     1       -4.506304 -4.038029 8669 2.695104e-05  0.11681929 \n",
       "5 -13.75724     1       -4.617702 -3.930953 8669 4.230485e-05  0.12224692 \n",
       "6 -13.56790     1       -4.196342 -4.057798 8669 2.476879e-05  0.13289494 \n",
       "  target_gene_region\n",
       "1 target            \n",
       "2 target            \n",
       "3 target            \n",
       "4 target            \n",
       "5 target            \n",
       "6 target            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(motif_targets_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ec3315dc-8b02-4b51-bbf1-9f87ed51e09f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining, by = \"representative_motif\"\n"
     ]
    }
   ],
   "source": [
    "motif_targets_genes <- motif_targets_genes %>%\n",
    "                       left_join(.,\n",
    "                                 rbp_motif_coefs_gex_sig %>% \n",
    "                                 mutate(RBP_GENENAME = sub(\"_.*\",\"\",GENENAME)) %>%\n",
    "                                 distinct(cancer,representative_motif,Motif_ID,RBP_GENENAME,log2FoldChange))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ebc0a685-03bd-4f1b-a766-bcf491928ae3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>event_type</th><th scope=col>window_size</th><th scope=col>seq_id</th><th scope=col>representative_motif</th><th scope=col>gene_id</th><th scope=col>gene_affinity</th><th scope=col>nevents</th><th scope=col>ma</th><th scope=col>zscore</th><th scope=col>n</th><th scope=col>zscore_pvalue</th><th scope=col>zscore_padj</th><th scope=col>target_gene_region</th><th scope=col>cancer</th><th scope=col>Motif_ID</th><th scope=col>log2FoldChange</th><th scope=col>RBP_GENENAME</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>BLCA</td><td>M012_0.6</td><td>-1.4714000</td><td>CPEB3</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>BRCA</td><td>M012_0.6</td><td>-0.1372552</td><td>CPEB3</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>CESC</td><td>M012_0.6</td><td>-0.4280138</td><td>CPEB3</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>CHOL</td><td>M012_0.6</td><td>-2.9633960</td><td>CPEB3</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>COAD</td><td>M012_0.6</td><td>-2.5881499</td><td>CPEB3</td></tr>\n",
       "\t<tr><td>A3</td><td>200</td><td>S3_DOWN_200</td><td>M012_0.6</td><td>ENSG00000104852</td><td>-14.17628</td><td>1</td><td>-4.617702</td><td>-4.111184</td><td>8669</td><td>1.968173e-05</td><td>0.08531045</td><td>target</td><td>ESCA</td><td>M012_0.6</td><td>-1.4573021</td><td>CPEB3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 17\n",
       "\\begin{tabular}{lllllllllllllllll}\n",
       " event\\_type & window\\_size & seq\\_id & representative\\_motif & gene\\_id & gene\\_affinity & nevents & ma & zscore & n & zscore\\_pvalue & zscore\\_padj & target\\_gene\\_region & cancer & Motif\\_ID & log2FoldChange & RBP\\_GENENAME\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & BLCA & M012\\_0.6 & -1.4714000 & CPEB3\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & BRCA & M012\\_0.6 & -0.1372552 & CPEB3\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & CESC & M012\\_0.6 & -0.4280138 & CPEB3\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & CHOL & M012\\_0.6 & -2.9633960 & CPEB3\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & COAD & M012\\_0.6 & -2.5881499 & CPEB3\\\\\n",
       "\t A3 & 200 & S3\\_DOWN\\_200 & M012\\_0.6 & ENSG00000104852 & -14.17628 & 1 & -4.617702 & -4.111184 & 8669 & 1.968173e-05 & 0.08531045 & target & ESCA & M012\\_0.6 & -1.4573021 & CPEB3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 17\n",
       "\n",
       "| event_type &lt;chr&gt; | window_size &lt;chr&gt; | seq_id &lt;chr&gt; | representative_motif &lt;chr&gt; | gene_id &lt;chr&gt; | gene_affinity &lt;dbl&gt; | nevents &lt;int&gt; | ma &lt;dbl&gt; | zscore &lt;dbl&gt; | n &lt;int&gt; | zscore_pvalue &lt;dbl&gt; | zscore_padj &lt;dbl&gt; | target_gene_region &lt;chr&gt; | cancer &lt;chr&gt; | Motif_ID &lt;chr&gt; | log2FoldChange &lt;dbl&gt; | RBP_GENENAME &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | BLCA | M012_0.6 | -1.4714000 | CPEB3 |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | BRCA | M012_0.6 | -0.1372552 | CPEB3 |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | CESC | M012_0.6 | -0.4280138 | CPEB3 |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | CHOL | M012_0.6 | -2.9633960 | CPEB3 |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | COAD | M012_0.6 | -2.5881499 | CPEB3 |\n",
       "| A3 | 200 | S3_DOWN_200 | M012_0.6 | ENSG00000104852 | -14.17628 | 1 | -4.617702 | -4.111184 | 8669 | 1.968173e-05 | 0.08531045 | target | ESCA | M012_0.6 | -1.4573021 | CPEB3 |\n",
       "\n"
      ],
      "text/plain": [
       "  event_type window_size seq_id      representative_motif gene_id        \n",
       "1 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "2 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "3 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "4 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "5 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "6 A3         200         S3_DOWN_200 M012_0.6             ENSG00000104852\n",
       "  gene_affinity nevents ma        zscore    n    zscore_pvalue zscore_padj\n",
       "1 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "2 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "3 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "4 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "5 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "6 -14.17628     1       -4.617702 -4.111184 8669 1.968173e-05  0.08531045 \n",
       "  target_gene_region cancer Motif_ID log2FoldChange RBP_GENENAME\n",
       "1 target             BLCA   M012_0.6 -1.4714000     CPEB3       \n",
       "2 target             BRCA   M012_0.6 -0.1372552     CPEB3       \n",
       "3 target             CESC   M012_0.6 -0.4280138     CPEB3       \n",
       "4 target             CHOL   M012_0.6 -2.9633960     CPEB3       \n",
       "5 target             COAD   M012_0.6 -2.5881499     CPEB3       \n",
       "6 target             ESCA   M012_0.6 -1.4573021     CPEB3       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(motif_targets_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7f2c2059-3637-4eee-9eb8-ca9c5bfa799a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'RBP_GENENAME'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 95 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>RBP_GENENAME</th><th scope=col>target_gene_region</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ACO1      </td><td>target</td><td>106</td></tr>\n",
       "\t<tr><td>ANKHD1    </td><td>target</td><td>  1</td></tr>\n",
       "\t<tr><td>ANKRD17   </td><td>target</td><td>  1</td></tr>\n",
       "\t<tr><td>CELF3     </td><td>target</td><td>138</td></tr>\n",
       "\t<tr><td>CELF4     </td><td>target</td><td>138</td></tr>\n",
       "\t<tr><td>CIRBP     </td><td>target</td><td>331</td></tr>\n",
       "\t<tr><td>CNOT4     </td><td>target</td><td> 14</td></tr>\n",
       "\t<tr><td>CPEB3     </td><td>target</td><td>106</td></tr>\n",
       "\t<tr><td>CPEB4     </td><td>target</td><td> 39</td></tr>\n",
       "\t<tr><td>DAZAP1    </td><td>target</td><td>289</td></tr>\n",
       "\t<tr><td>EIF2S1    </td><td>target</td><td>  2</td></tr>\n",
       "\t<tr><td>EIF4B     </td><td>target</td><td>428</td></tr>\n",
       "\t<tr><td>ELAVL1    </td><td>target</td><td>  8</td></tr>\n",
       "\t<tr><td>ELAVL2    </td><td>target</td><td>367</td></tr>\n",
       "\t<tr><td>ELAVL3    </td><td>target</td><td>360</td></tr>\n",
       "\t<tr><td>ENOX1     </td><td>target</td><td> 14</td></tr>\n",
       "\t<tr><td>ENOX2     </td><td>target</td><td> 14</td></tr>\n",
       "\t<tr><td>FMR1      </td><td>target</td><td>119</td></tr>\n",
       "\t<tr><td>FXR2      </td><td>target</td><td>119</td></tr>\n",
       "\t<tr><td>G3BP2     </td><td>target</td><td>178</td></tr>\n",
       "\t<tr><td>HNRNPA1   </td><td>target</td><td> 76</td></tr>\n",
       "\t<tr><td>HNRNPA1L2 </td><td>target</td><td> 76</td></tr>\n",
       "\t<tr><td>HNRNPA1P36</td><td>target</td><td> 76</td></tr>\n",
       "\t<tr><td>HNRNPA1P7 </td><td>target</td><td> 97</td></tr>\n",
       "\t<tr><td>HNRNPA3   </td><td>target</td><td> 76</td></tr>\n",
       "\t<tr><td>HNRNPC    </td><td>target</td><td> 39</td></tr>\n",
       "\t<tr><td>HNRNPDL   </td><td>target</td><td>289</td></tr>\n",
       "\t<tr><td>HNRNPK    </td><td>target</td><td>  6</td></tr>\n",
       "\t<tr><td>HNRNPL    </td><td>target</td><td>  3</td></tr>\n",
       "\t<tr><td>HNRNPR    </td><td>target</td><td> 68</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>RBMS1  </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>RBMS2  </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>RBMS2P1</td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>RBMS3  </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>RBMXL1 </td><td>target</td><td> 215</td></tr>\n",
       "\t<tr><td>RBMXL2 </td><td>target</td><td> 215</td></tr>\n",
       "\t<tr><td>SFPQ   </td><td>target</td><td>  95</td></tr>\n",
       "\t<tr><td>SNRNP70</td><td>target</td><td> 108</td></tr>\n",
       "\t<tr><td>SNRPA  </td><td>target</td><td>1780</td></tr>\n",
       "\t<tr><td>SNRPB2 </td><td>target</td><td>1081</td></tr>\n",
       "\t<tr><td>SRSF1  </td><td>target</td><td> 503</td></tr>\n",
       "\t<tr><td>SRSF10 </td><td>target</td><td>1321</td></tr>\n",
       "\t<tr><td>SRSF12 </td><td>target</td><td>1321</td></tr>\n",
       "\t<tr><td>SRSF2  </td><td>target</td><td> 639</td></tr>\n",
       "\t<tr><td>SRSF3  </td><td>target</td><td>  29</td></tr>\n",
       "\t<tr><td>SRSF4  </td><td>target</td><td>  34</td></tr>\n",
       "\t<tr><td>SRSF6  </td><td>target</td><td>  34</td></tr>\n",
       "\t<tr><td>SRSF7  </td><td>target</td><td>  25</td></tr>\n",
       "\t<tr><td>SRSF9  </td><td>target</td><td>  60</td></tr>\n",
       "\t<tr><td>SYNCRIP</td><td>target</td><td>  68</td></tr>\n",
       "\t<tr><td>TARDBP </td><td>target</td><td> 178</td></tr>\n",
       "\t<tr><td>TIA1   </td><td>target</td><td>  39</td></tr>\n",
       "\t<tr><td>TUT1   </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>YBX1   </td><td>target</td><td>  29</td></tr>\n",
       "\t<tr><td>YBX2   </td><td>target</td><td> 422</td></tr>\n",
       "\t<tr><td>YBX3   </td><td>target</td><td> 435</td></tr>\n",
       "\t<tr><td>ZC3H10 </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>ZFP36  </td><td>target</td><td>   6</td></tr>\n",
       "\t<tr><td>ZFP36L1</td><td>target</td><td> 356</td></tr>\n",
       "\t<tr><td>ZFP36L2</td><td>target</td><td> 356</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 95 × 3\n",
       "\\begin{tabular}{lll}\n",
       " RBP\\_GENENAME & target\\_gene\\_region & n\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t ACO1       & target & 106\\\\\n",
       "\t ANKHD1     & target &   1\\\\\n",
       "\t ANKRD17    & target &   1\\\\\n",
       "\t CELF3      & target & 138\\\\\n",
       "\t CELF4      & target & 138\\\\\n",
       "\t CIRBP      & target & 331\\\\\n",
       "\t CNOT4      & target &  14\\\\\n",
       "\t CPEB3      & target & 106\\\\\n",
       "\t CPEB4      & target &  39\\\\\n",
       "\t DAZAP1     & target & 289\\\\\n",
       "\t EIF2S1     & target &   2\\\\\n",
       "\t EIF4B      & target & 428\\\\\n",
       "\t ELAVL1     & target &   8\\\\\n",
       "\t ELAVL2     & target & 367\\\\\n",
       "\t ELAVL3     & target & 360\\\\\n",
       "\t ENOX1      & target &  14\\\\\n",
       "\t ENOX2      & target &  14\\\\\n",
       "\t FMR1       & target & 119\\\\\n",
       "\t FXR2       & target & 119\\\\\n",
       "\t G3BP2      & target & 178\\\\\n",
       "\t HNRNPA1    & target &  76\\\\\n",
       "\t HNRNPA1L2  & target &  76\\\\\n",
       "\t HNRNPA1P36 & target &  76\\\\\n",
       "\t HNRNPA1P7  & target &  97\\\\\n",
       "\t HNRNPA3    & target &  76\\\\\n",
       "\t HNRNPC     & target &  39\\\\\n",
       "\t HNRNPDL    & target & 289\\\\\n",
       "\t HNRNPK     & target &   6\\\\\n",
       "\t HNRNPL     & target &   3\\\\\n",
       "\t HNRNPR     & target &  68\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t RBMS1   & target &    6\\\\\n",
       "\t RBMS2   & target &    6\\\\\n",
       "\t RBMS2P1 & target &    6\\\\\n",
       "\t RBMS3   & target &    6\\\\\n",
       "\t RBMXL1  & target &  215\\\\\n",
       "\t RBMXL2  & target &  215\\\\\n",
       "\t SFPQ    & target &   95\\\\\n",
       "\t SNRNP70 & target &  108\\\\\n",
       "\t SNRPA   & target & 1780\\\\\n",
       "\t SNRPB2  & target & 1081\\\\\n",
       "\t SRSF1   & target &  503\\\\\n",
       "\t SRSF10  & target & 1321\\\\\n",
       "\t SRSF12  & target & 1321\\\\\n",
       "\t SRSF2   & target &  639\\\\\n",
       "\t SRSF3   & target &   29\\\\\n",
       "\t SRSF4   & target &   34\\\\\n",
       "\t SRSF6   & target &   34\\\\\n",
       "\t SRSF7   & target &   25\\\\\n",
       "\t SRSF9   & target &   60\\\\\n",
       "\t SYNCRIP & target &   68\\\\\n",
       "\t TARDBP  & target &  178\\\\\n",
       "\t TIA1    & target &   39\\\\\n",
       "\t TUT1    & target &    6\\\\\n",
       "\t YBX1    & target &   29\\\\\n",
       "\t YBX2    & target &  422\\\\\n",
       "\t YBX3    & target &  435\\\\\n",
       "\t ZC3H10  & target &    6\\\\\n",
       "\t ZFP36   & target &    6\\\\\n",
       "\t ZFP36L1 & target &  356\\\\\n",
       "\t ZFP36L2 & target &  356\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 95 × 3\n",
       "\n",
       "| RBP_GENENAME &lt;chr&gt; | target_gene_region &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| ACO1       | target | 106 |\n",
       "| ANKHD1     | target |   1 |\n",
       "| ANKRD17    | target |   1 |\n",
       "| CELF3      | target | 138 |\n",
       "| CELF4      | target | 138 |\n",
       "| CIRBP      | target | 331 |\n",
       "| CNOT4      | target |  14 |\n",
       "| CPEB3      | target | 106 |\n",
       "| CPEB4      | target |  39 |\n",
       "| DAZAP1     | target | 289 |\n",
       "| EIF2S1     | target |   2 |\n",
       "| EIF4B      | target | 428 |\n",
       "| ELAVL1     | target |   8 |\n",
       "| ELAVL2     | target | 367 |\n",
       "| ELAVL3     | target | 360 |\n",
       "| ENOX1      | target |  14 |\n",
       "| ENOX2      | target |  14 |\n",
       "| FMR1       | target | 119 |\n",
       "| FXR2       | target | 119 |\n",
       "| G3BP2      | target | 178 |\n",
       "| HNRNPA1    | target |  76 |\n",
       "| HNRNPA1L2  | target |  76 |\n",
       "| HNRNPA1P36 | target |  76 |\n",
       "| HNRNPA1P7  | target |  97 |\n",
       "| HNRNPA3    | target |  76 |\n",
       "| HNRNPC     | target |  39 |\n",
       "| HNRNPDL    | target | 289 |\n",
       "| HNRNPK     | target |   6 |\n",
       "| HNRNPL     | target |   3 |\n",
       "| HNRNPR     | target |  68 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| RBMS1   | target |    6 |\n",
       "| RBMS2   | target |    6 |\n",
       "| RBMS2P1 | target |    6 |\n",
       "| RBMS3   | target |    6 |\n",
       "| RBMXL1  | target |  215 |\n",
       "| RBMXL2  | target |  215 |\n",
       "| SFPQ    | target |   95 |\n",
       "| SNRNP70 | target |  108 |\n",
       "| SNRPA   | target | 1780 |\n",
       "| SNRPB2  | target | 1081 |\n",
       "| SRSF1   | target |  503 |\n",
       "| SRSF10  | target | 1321 |\n",
       "| SRSF12  | target | 1321 |\n",
       "| SRSF2   | target |  639 |\n",
       "| SRSF3   | target |   29 |\n",
       "| SRSF4   | target |   34 |\n",
       "| SRSF6   | target |   34 |\n",
       "| SRSF7   | target |   25 |\n",
       "| SRSF9   | target |   60 |\n",
       "| SYNCRIP | target |   68 |\n",
       "| TARDBP  | target |  178 |\n",
       "| TIA1    | target |   39 |\n",
       "| TUT1    | target |    6 |\n",
       "| YBX1    | target |   29 |\n",
       "| YBX2    | target |  422 |\n",
       "| YBX3    | target |  435 |\n",
       "| ZC3H10  | target |    6 |\n",
       "| ZFP36   | target |    6 |\n",
       "| ZFP36L1 | target |  356 |\n",
       "| ZFP36L2 | target |  356 |\n",
       "\n"
      ],
      "text/plain": [
       "   RBP_GENENAME target_gene_region n   \n",
       "1  ACO1         target             106 \n",
       "2  ANKHD1       target               1 \n",
       "3  ANKRD17      target               1 \n",
       "4  CELF3        target             138 \n",
       "5  CELF4        target             138 \n",
       "6  CIRBP        target             331 \n",
       "7  CNOT4        target              14 \n",
       "8  CPEB3        target             106 \n",
       "9  CPEB4        target              39 \n",
       "10 DAZAP1       target             289 \n",
       "11 EIF2S1       target               2 \n",
       "12 EIF4B        target             428 \n",
       "13 ELAVL1       target               8 \n",
       "14 ELAVL2       target             367 \n",
       "15 ELAVL3       target             360 \n",
       "16 ENOX1        target              14 \n",
       "17 ENOX2        target              14 \n",
       "18 FMR1         target             119 \n",
       "19 FXR2         target             119 \n",
       "20 G3BP2        target             178 \n",
       "21 HNRNPA1      target              76 \n",
       "22 HNRNPA1L2    target              76 \n",
       "23 HNRNPA1P36   target              76 \n",
       "24 HNRNPA1P7    target              97 \n",
       "25 HNRNPA3      target              76 \n",
       "26 HNRNPC       target              39 \n",
       "27 HNRNPDL      target             289 \n",
       "28 HNRNPK       target               6 \n",
       "29 HNRNPL       target               3 \n",
       "30 HNRNPR       target              68 \n",
       "⋮  ⋮            ⋮                  ⋮   \n",
       "66 RBMS1        target                6\n",
       "67 RBMS2        target                6\n",
       "68 RBMS2P1      target                6\n",
       "69 RBMS3        target                6\n",
       "70 RBMXL1       target              215\n",
       "71 RBMXL2       target              215\n",
       "72 SFPQ         target               95\n",
       "73 SNRNP70      target              108\n",
       "74 SNRPA        target             1780\n",
       "75 SNRPB2       target             1081\n",
       "76 SRSF1        target              503\n",
       "77 SRSF10       target             1321\n",
       "78 SRSF12       target             1321\n",
       "79 SRSF2        target              639\n",
       "80 SRSF3        target               29\n",
       "81 SRSF4        target               34\n",
       "82 SRSF6        target               34\n",
       "83 SRSF7        target               25\n",
       "84 SRSF9        target               60\n",
       "85 SYNCRIP      target               68\n",
       "86 TARDBP       target              178\n",
       "87 TIA1         target               39\n",
       "88 TUT1         target                6\n",
       "89 YBX1         target               29\n",
       "90 YBX2         target              422\n",
       "91 YBX3         target              435\n",
       "92 ZC3H10       target                6\n",
       "93 ZFP36        target                6\n",
       "94 ZFP36L1      target              356\n",
       "95 ZFP36L2      target              356"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "motif_targets_genes  %>%\n",
    "    group_by(RBP_GENENAME,target_gene_region) %>%\n",
    "    summarize(n=length(unique(gene_id))) %>%\n",
    "    filter(target_gene_region==\"target\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3e18f1bf-dc04-45f8-aa5a-183055b8e52e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining, by = c(\"event_type\", \"seq_id\", \"representative_motif\", \"cancer\",\n",
      "\"Motif_ID\", \"log2FoldChange\")\n"
     ]
    }
   ],
   "source": [
    "rbp.targets <- inner_join(ungroup(motif_targets_genes),\n",
    "                         rbp_motif_coefs_gex_sig %>% \n",
    "                         mutate(seq_id = paste(position,direction,wsize,sep=\"_\"))) %>%\n",
    "                 mutate(RBP_GENENAME=sub(\"_.*\",\"\",GENENAME),\n",
    "                        RBP_dgex_group = ifelse(log2FoldChange>0,\"up\",\"down\")) %>%\n",
    "                 rename(\"RBP.log2FoldChange\"=\"log2FoldChange\")  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13a0c7e-90cf-4f0f-a391-ee35c17b0360",
   "metadata": {},
   "source": [
    "## Save objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e651def3-a3c4-489e-8825-8c26680be026",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(motif_targets_genes,file=\"../output/objects/Figure10_all_motif_targets.RDS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "40da92bc-54fe-4433-a62f-336f7a6af324",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(rbp.targets,file=\"../output/objects/Figure10_rbp.targets.RDS\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
